Added SSE support for real-time updates, improved Docker / Portainer deployment, OpenHab config samples, and a few minor improvements

JonathanGiles opened the pull request:
This PR adds improved support for deploying via Docker Compose / Portainer, it adds some sample OpenHab config files, it adds a 'PUBLISH_INTERVAL' property to reduce the frequency of publishing (whilst retaining the current approach of firing all events as soon as they are received). 

I also took the liberty of tidying up the markdown files and splitting them out into different topics. Let me know if you want to keep it like that. In my mind it makes it easier to go to the topic you are interested in without getting overwhelmed or lost in unrelated content.

I tested this on my home Enphase system, running on version D8.3.5232 (f18f5b).

JonathanGiles committed:
Fixes for docker / portainer, and some misc improvements.

JonathanGiles committed:
More cleanup

JonathanGiles committed:
markdown cleanup

JonathanGiles committed:
fix title

JonathanGiles committed:
Replace credentials with placeholders in options.json

vk2him commented:
Thanks for the work Jonathan, it's a busy time of the year so I won't have a chance to do anything until the new year - cheers

vk2him commented:
@JonathanGiles  - a few changes to the base code were made in this recent fix https://github.com/vk2him/Enphase-Envoy-mqtt-json/issues/48  - would you mind checking your pull request to see if it needs to be modified as a result of the recent change? Thanks

JonathanGiles committed:
Merge upstream/main: resolve conflicts with HTTPS option and boolean fixes

JonathanGiles committed:
Add SSE streaming support for v7/v8 firmware

- Add USE_SSE config option (default: true) to use /stream/meter SSE endpoint
- Add scrape_stream_sse() for token-auth firmware (v7/v8)
- Wire SSE into main() with fallback to polling when USE_SSE is false
- Update config.yaml, .env.example, docker-compose.yml, options.json

JonathanGiles committed:
Fix missing ENVOY_USE_HTTPS in env vars config path

JonathanGiles committed:
Add SSE/polling example folders and update docs

- Create examples/sse/ with Home Assistant and OpenHAB configs for SSE mode
- Create examples/polling/ with Home Assistant and OpenHAB configs for polling mode
- Add examples/README.md explaining both modes and JSON structures
- Update HOME-ASSISTANT.md to reference examples/ directory
- Update openhab/README_OPENHAB.md with SSE vs polling callout
- Add SSE-MIGRATION.md migration guide
- Update README.md with SSE documentation

JonathanGiles committed:
Add March 2026 note for SSE support in README and CHANGELOG

JonathanGiles committed:
Remove SSE-MIGRATION.md (content covered by examples/README.md)

JonathanGiles commented:
@vk2him I have added SSE support and fixed this up to merge again. The SSE support means i can get real-time updates from my enphase system without overwhelming it (because it is pushing to me, rather than me polling).

vk2him commented:
Hi Jonathan,
I'm very grateful for the effort you have put into these changes, it will be a huge assistance for those that want to use these additions.

I have a concern however as I have an older Enphase system with v5 firmware, and I have a very limited use case of feeding the enphase json output from /stream/meter to mqtt.  Therefore, I'm only able to test the portions of the code that relate to v5, and if I accept your merge request I'll effectively be unable to confirm/fault-find/test any issues that may arise that uses > v5 firmware and the additional improvements your code brings.

As a suggestion, rather that merge your changes into my code, what would you think if I added a section at the front of my readme.md pointing to your repository if anyone wanted to utilise your new features? That way my code would stay relatively stationary, with changes only made by myself to resolve any potential breaking changes as a result of updated/deprecated functions?

Many thanks
Ian

JonathanGiles commented:
That's perfectly fine.

sddgit commented:
I'm confused! Hasn't the add-on been using a steam (push) point all along? I know I must be missing something...

JonathanGiles commented:
> I'm confused! Hasn't the add-on been using a steam (push) point all along? I know I must be missing something...

on v5 it uses the stream endpoint, but on later versions it was changed to only read from a polling endpoint. This code change allows for the stream endpoint to be read (using the new token auth) on newer releases. For my use case (a home kiosk) where I wanted real-time readings, the SSE endpoint was my preferred approach so I didn't have to either overwhelm the endpoint or poll every 30s or so.

sddgit commented:
I'm on V8 and receiving 1-second updates, just as it did at a different house using V5. So is it now polling every second by default? Does your new code result in the same JSON as I'm getting currently? The templates I’ve developed are at https://github.com/vk2him/Enphase-Envoy-mqtt-json/issues/42.

JonathanGiles commented:
Given you are on version 8, I believe (I am not a software engineer, I just play one on TV, etc) you will start off [here on line 443](https://github.com/vk2him/Enphase-Envoy-mqtt-json/blob/main/envoy_to_mqtt_json.py#L443), which then takes you to `scrape_stream_meters` [line 363](https://github.com/vk2him/Enphase-Envoy-mqtt-json/blob/main/envoy_to_mqtt_json.py#L363), which as you can see, calls to your local envoy and goes to `/ivp/meters/readings` in the URL path. That, as far as I understand, is a polling endpoint (you can read the details [here](https://enphase.com/download/iq-gateway-local-apis-or-ui-access-using-token)). In a past life I used to crash my enphase envoy system by polling too regularly, so I always used the server-sent events endpoint, as then it was up to the envoy system to push to me instead.

You can see the existing code base does more or less this already for v5 [here](https://github.com/vk2him/Enphase-Envoy-mqtt-json/blob/main/envoy_to_mqtt_json.py#L437), by going into the `scrape_stream` path [here](https://github.com/vk2him/Enphase-Envoy-mqtt-json/blob/main/envoy_to_mqtt_json.py#L399). This path just pushes the SSE events to the MQTT topic whenever they are received (which, is about once every second).

The JSON is different, but that is why my PR here covers both JSON outputs. I used AI to transfer my Home Kiosk system from the old JSON to the new JSON - it took about 2 seconds.

sddgit commented:
Looking at https://github.com/vk2him/Enphase-Envoy-mqtt-json/issues/15#issuecomment-1641906709, it's stated that the addon uses a stream even after V7, and only provides 1-second updates, so I'm still confused!

JonathanGiles commented:
It's hard to know what is being referred to in that comment. It could be referring to the original v5 code that is definitely still using the stream pathway. That would be my guess.

vk2him commented:
> It's hard to know what is being referred to in that comment. It could be referring to the original v5 code that is definitely still using the stream pathway. That would be my guess.

Hi guys - sorry for the confusion in all this.
Yes in the context of that comment at the time it was made, I was referring to the v5 stream end point http://envoy.local/stream/meter as that is a stream - all other end points are not streams and you need to poll them manually for every updated reading.

So @sddgit   - the v8 end point I assum you're using is http://envoy.local/ivp/meters/readings and that's not a stream and needs to be polled every second which is what the code does and requires a different template to extract the resulting values.

@JonathanGiles has made a change that allows the envoy.local/stream/meter end point, which is a stream and doesn't need polling, to be used for all versions of firmware as currently, it's only used if you have v5 firmware. Note this will require a different template as the values are in different locations.

In the end, the polling end point and the stream endpoint give 1 second resolution. The question is if the polling end point puts additional load on the Envoy (causing it to lock up) compared to the streaming endpoint, and that seems to be at least one of the reasons why @JonathanGiles made the changes as he saw issues with his Envoy when polled regularly. I haven't seen any issues raised regarding polling every second, and when I first developed this code, I initally used the polling end point for months without issues before I found the streaming endpoint and I then published this repository. 

Edited to add:

I could simply add an optional override that forced all firmware versions to use envoy.local/stream/meter if people wanted to use the stream output rather than the polling method. The default would be 'false' so as to not interfere with any existing users since a template change would probably be required . Does anyone have any input on that? Thanks

sddgit commented:
Thanks for clearing that up @vk2him. I had no idea polling is used for later versions! But, as you experienced, I’ve seen no issues to date. I can’t see stream/meter mentioned in the documentation cited above @JonathanGiles. Is that endpoint something that somebody has just tried more recently, since I'm pretty sure it didn't exist when Enphase first made the change to tokens.

I do think the PR is a very valuable addition. If I may play devil's advocate, accepting this change would seem to me to be no different to accepting the major changes that were involved for V7/8. You couldn’t test those either, but the community worked together to get things implemented. I expect SSE couldn’t be the default though, since it breaks existing templates. Still, you're the boss if you think otherwise!

JonathanGiles commented:
> Edited to add:
> 
> I could simply add an optional override that forced all firmware versions to use envoy.local/stream/meter if people wanted to use the stream output rather than the polling method. The default would be 'false' so as to not interfere with any existing users since a template change would probably be required . Does anyone have any input on that? Thanks

From a purity standpoint, it just feels better to have the envoy pushing at a pace it can sustain rather than polling it regardless of the load it is under. Like I said, in my last house I used to experience issues with the system not responding well - but I was polling multiple endpoints for the various details at various URLs. I found the SSE endpoint far more reliable - but clearly this isn't your lived experience. I still feel better about SSE over polling though.

JonathanGiles commented:
> I can’t see stream/meter mentioned in the documentation cited above @JonathanGiles. Is that endpoint something that somebody has just tried more recently, since I'm pretty sure it didn't exist when Enphase first made the change to tokens.

/stream/meter has been around for quite a long time. I used it at my last house since ~2018 or so. Like I said, I'm using it today on version 8 of the envoy software at my current house.

vk2him commented:
> I can’t see stream/meter mentioned in the documentation cited above @JonathanGiles. Is that endpoint something that somebody has just tried more recently, since I'm pretty sure it didn't exist when Enphase first made the change to tokens.

@sddgit  - https://envoy.local/stream/meter (or https://envoy.local/stream/meter) have always been available, it is currently only used for v5 firmware.

As a test, try navigating to https://envoy.local/stream/meter - the username is installer and you'll need the password from passwordCalc.py in the repository ... As I don't have a v7/v8 system I can't test, I don't think you need an enphase token but I may be wrong

vk2him commented:
> From a purity standpoint, it just feels better to have the envoy pushing at a pace it can sustain rather than polling it regardless of the load it is under. Like I said, in my last house I used to experience issues with the system not responding well - but I was polling multiple endpoints for the various details at various URLs. I found the SSE endpoint far more reliable - but clearly this isn't your lived experience. I still feel better about SSE over polling though.

Perhaps polling multiple end-points in parallel may cause an issue - I've never tried but I could see it potentially causing memory leaks. Does https://envoy.local/stream/meter require a token on v7/v8 or just the installer password?

JonathanGiles commented:
I'm out but I'm fairly certain it requires the token.

sddgit commented:
I agree that steaming rather than polling is much better from a purity standpoint. I do seem to remember (from the very long and angry post on the Enphase forum when tokens first came in) that the streaming endpoint was removed but maybe came back later. I could be wrong though.

From your comment about making the stream optional @vk2him (which would make sense), do I take it that you're considering using this PR?

vk2him commented:
@sddgit - yes I'm considering but I need to know if the streaming end point it needs a token on a v8 system, or just the password. See my earlier comment where I ask if you could please test? Thanks

sddgit commented:
I went to https://envoy.local/stream/meter, and got directed to the Envoy page requesting a token. Generated one and entered it and it proceeded to the standard Envoy page. Then went to https://envoy.local/stream/meter again and it accessed the stream.

BTW, if that endpoint existed all along, does anyone know why a polling one was used instead for V7?

JonathanGiles commented:
When I looked earlier (I'm out on my phone now), my conclusion was that the newer version support was added by someone else... And that is what they decided to add.

sddgit commented:
Yes, it was definitely added by someone else. Makes me think that my memory is correct and the streaming endpoint disappeared for a time there...

vk2him commented:
Thanks everyone for your discussion points - the version 7/8 code sections were added as a Pull request a few years ago by @helderd per the Readme file. I don't have a v7/8 system so before I approved the merge, a number of users tested it.

Now  I know that the streaming endpoint requires a token on > v5 firmware, it makes sense for me to merge @JonathanGiles request into the main.

I do have a caveat however - making SSE ```true```   by default would be  abreaking change for all non V5 users as they will need to use the "old v5 templates".  

So @JonathanGiles , if you agree can you please make two changes?

1. Please set SSE option to ```false```    in all areas
2. Make a change on line 8 of the README.md file to say words to the effect of "SSE is the suggested default method to be used. By default it is set to false as existing users will need to use modified templates. Step by step assistance is provided below to assist you in creating these templates if you decide to use the SSE method.

After all the changes are made, I'll merge but I won't bump the version number in ```config.yaml``` until we jointly test it first - I don't want to alert all Home Assistant users of a new version until we test it. If you're not aware, if you go to my addon in your home assistant (I think they now call it an App), if you press the Red ```Rebuild``` button, it will download the latest copy of the code from Github - this is how we can test it before we publish it to the HA community.

Again, many thanks to you all for this, regards Ian

JonathanGiles committed:
changing sse to not be default

JonathanGiles commented:
Changes made as requested.

sddgit commented:
@JonathanGiles Can you help me with how you got AI to change your templates to reflect the streaming endpoint's JSON? What prompt did you use? As I said above, I use a large collection of templates, as in https://github.com/vk2him/Enphase-Envoy-mqtt-json/issues/42. I'm hoping they will all still work once converted.

JonathanGiles commented:
I just gave the AI the JSON as it was from the old polling endpoints and the new SSE endpoint, and told it to update my code. It happened easily after that.

sddgit commented:
Sorry, when you say "gave the AI the JSON", what exactly did you give it? Where do I find the complete JSON it would need for both cases? Posting here would be useful I think.

JonathanGiles commented:
Look at all the files in my PR.

sddgit commented:
OK, this is what I have for a set of 3-phase sensors for SSE mode. I'd be grateful if people could cast a quick eye over them. The polling sensors have overall values for production, consumption etc as well as per-phase. The streaming ones don't have this, so this set creates them. Polling is also net consumption, whereas streaming is total, so no need for calculations any more.

I have noticed that in the streaming homeassistant.yaml example it still says SSE is the default. Also says that in the README.

If this all looks good, I'm ready to test a merged new version of the add-on. Thanks all for the great work.

```
# ── Net ─────────────────────────────────────────────────────
- name: "mqtt_net"
  unique_id: "envoy_mqtt_net"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: >
    {{ (
      value_json["net-consumption"]["ph-a"]["p"] +
      value_json["net-consumption"]["ph-b"]["p"] +
      value_json["net-consumption"]["ph-c"]["p"]
    ) | round(0) | int(0) }}

- name: "mqtt_net_phase_a"
  unique_id: "envoy_mqtt_net_phase_a"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["net-consumption"]["ph-a"]["p"] | round(0) | int(0) }}'

- name: "mqtt_net_phase_b"
  unique_id: "envoy_mqtt_net_phase_b"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["net-consumption"]["ph-b"]["p"] | round(0) | int(0) }}'

- name: "mqtt_net_phase_c"
  unique_id: "envoy_mqtt_net_phase_c"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["net-consumption"]["ph-c"]["p"] | round(0) | int(0) }}'

# ── Production ──────────────────────────────────────────────
- name: "mqtt_production"
  unique_id: "envoy_mqtt_production"
  state_topic: "/envoy/json"
  qos: 0
  unit_of_measurement: "W"
  value_template: >
    {% set v = (
      value_json["production"]["ph-a"]["p"] +
      value_json["production"]["ph-b"]["p"] +
      value_json["production"]["ph-c"]["p"]
    ) | round(0) | int(0) %}
    {% if v <= 5 %}
      0
    {% elif is_state('sun.sun', 'below_horizon') %}
      0
    {% else %}
      {{ v }}
    {% endif %}
  state_class: measurement
  device_class: power

- name: "mqtt_production_ph_a"
  unique_id: "envoy_mqtt_production_ph_a"
  state_topic: "/envoy/json"
  qos: 0
  unit_of_measurement: "W"
  value_template: '{{ value_json["production"]["ph-a"]["p"] | round(0) | int(0) }}'
  state_class: measurement
  device_class: power

- name: "mqtt_production_ph_b"
  unique_id: "envoy_mqtt_production_ph_b"
  state_topic: "/envoy/json"
  qos: 0
  unit_of_measurement: "W"
  value_template: '{{ value_json["production"]["ph-b"]["p"] | round(0) | int(0) }}'
  state_class: measurement
  device_class: power

- name: "mqtt_production_ph_c"
  unique_id: "envoy_mqtt_production_ph_c"
  state_topic: "/envoy/json"
  qos: 0
  unit_of_measurement: "W"
  value_template: '{{ value_json["production"]["ph-c"]["p"] | round(0) | int(0) }}'
  state_class: measurement
  device_class: power

# ── Total Consumption ───────────────────────────────────────
- name: "mqtt_consumption"
  unique_id: "envoy_mqtt_consumption"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: >
    {{ (
      value_json["total-consumption"]["ph-a"]["p"] +
      value_json["total-consumption"]["ph-b"]["p"] +
      value_json["total-consumption"]["ph-c"]["p"]
    ) | round(0) | int(0) }}

- name: "mqtt_consumption_ph_a"
  unique_id: "envoy_mqtt_consumption_ph_a"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["total-consumption"]["ph-a"]["p"] | round(0) | int(0) }}'

- name: "mqtt_consumption_ph_b"
  unique_id: "envoy_mqtt_consumption_ph_b"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["total-consumption"]["ph-b"]["p"] | round(0) | int(0) }}'

- name: "mqtt_consumption_ph_c"
  unique_id: "envoy_mqtt_consumption_ph_c"
  state_topic: "/envoy/json"
  unit_of_measurement: "W"
  qos: 0
  state_class: measurement
  device_class: power
  value_template: '{{ value_json["total-consumption"]["ph-c"]["p"] | round(0) | int(0) }}'

# ── Power Factor ────────────────────────────────────────────
- name: "mqtt_power_factor"
  unique_id: "envoy_mqtt_power_factor"
  state_topic: "/envoy/json"
  value_template: >
    {{ (
      value_json["total-consumption"]["ph-a"]["pf"] +
      value_json["total-consumption"]["ph-b"]["pf"] +
      value_json["total-consumption"]["ph-c"]["pf"]
    ) / 3 | round(2) }}
  qos: 0
  unit_of_measurement: "%"
  state_class: measurement
  device_class: power_factor

- name: "mqtt_power_factor_ph_a"
  unique_id: "envoy_mqtt_power_factor_ph_a"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-a"]["pf"] | round(2) }}'
  qos: 0
  unit_of_measurement: "%"
  state_class: measurement
  device_class: power_factor

- name: "mqtt_power_factor_ph_b"
  unique_id: "envoy_mqtt_power_factor_ph_b"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-b"]["pf"] | round(2) }}'
  qos: 0
  unit_of_measurement: "%"
  state_class: measurement
  device_class: power_factor

- name: "mqtt_power_factor_ph_c"
  unique_id: "envoy_mqtt_power_factor_ph_c"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-c"]["pf"] | round(2) }}'
  qos: 0
  unit_of_measurement: "%"
  state_class: measurement
  device_class: power_factor

# ── Voltage ─────────────────────────────────────────────────
- name: "mqtt_voltage"
  unique_id: "envoy_mqtt_voltage"
  state_topic: "/envoy/json"
  value_template: >
    {{ (
      value_json["total-consumption"]["ph-a"]["v"] +
      value_json["total-consumption"]["ph-b"]["v"] +
      value_json["total-consumption"]["ph-c"]["v"]
    ) / 3 | round(0) | int(0) }}
  qos: 0
  unit_of_measurement: "V"
  state_class: measurement
  device_class: voltage

- name: "mqtt_voltage_ph_a"
  unique_id: "envoy_mqtt_voltage_ph_a"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-a"]["v"] | round(0) | int(0) }}'
  qos: 0
  unit_of_measurement: "V"
  state_class: measurement
  device_class: voltage

- name: "mqtt_voltage_ph_b"
  unique_id: "envoy_mqtt_voltage_ph_b"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-b"]["v"] | round(0) | int(0) }}'
  qos: 0
  unit_of_measurement: "V"
  state_class: measurement
  device_class: voltage

- name: "mqtt_voltage_ph_c"
  unique_id: "envoy_mqtt_voltage_ph_c"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-c"]["v"] | round(0) | int(0) }}'
  qos: 0
  unit_of_measurement: "V"
  state_class: measurement
  device_class: voltage

# ── Current ─────────────────────────────────────────────────
- name: "mqtt_current"
  unique_id: "envoy_mqtt_current"
  state_topic: "/envoy/json"
  value_template: >
    {{ (
      value_json["total-consumption"]["ph-a"]["i"] +
      value_json["total-consumption"]["ph-b"]["i"] +
      value_json["total-consumption"]["ph-c"]["i"]
    ) | round(2) }}
  qos: 0
  unit_of_measurement: "A"
  state_class: measurement
  device_class: current

- name: "mqtt_current_ph_a"
  unique_id: "envoy_mqtt_current_ph_a"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-a"]["i"] | round(2) }}'
  qos: 0
  unit_of_measurement: "A"
  state_class: measurement
  device_class: current

- name: "mqtt_current_ph_b"
  unique_id: "envoy_mqtt_current_ph_b"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-b"]["i"] | round(2) }}'
  qos: 0
  unit_of_measurement: "A"
  state_class: measurement
  device_class: current

- name: "mqtt_current_ph_c"
  unique_id: "envoy_mqtt_current_ph_c"
  state_topic: "/envoy/json"
  value_template: '{{ value_json["total-consumption"]["ph-c"]["i"] | round(2) }}'
  qos: 0
  unit_of_measurement: "A"
  state_class: measurement
  device_class: current
```

vk2him commented:
@JonathanGiles @sddgit 

I approved and merged your pull request, however I was unable to update the Home Assistant App by the rebuild method I mentioned above as it failed with the following error that looks liek it didn't like the ```apk``` command line?

Do you have any insights as I have reverted the PR back to ensure the repository continues to operate. You might need to create a new PR with any mods you make?

```
2026-03-16 11:31:24.425 INFO (MainThread) [supervisor.docker.manager] Removing image 13e68335/amd64-addon-envoy_to_mqtt_json with latest
2026-03-16 11:31:24.426 INFO (MainThread) [supervisor.docker.manager] Removing image 13e68335/amd64-addon-envoy_to_mqtt_json with 1.0.22
2026-03-16 11:31:24.431 INFO (MainThread) [supervisor.docker.addon] Starting build for 13e68335/amd64-addon-envoy_to_mqtt_json:1.0.22
2026-03-16 11:31:24.432 INFO (MainThread) [supervisor.docker.manager] Running command '['docker', 'buildx', 'build', '.', '--tag', '13e68335/amd64-addon-envoy_to_mqtt_json:1.0.22', '--file', 'Dockerfile', '--platform', 'linux/amd64', '--pull', '--label', 'io.hass.version=1.0.22', '--label', 'io.hass.arch=amd64', '--label', 'io.hass.type=addon', '--label', 'io.hass.name=Stream mqtt from Enphase Envoy', '--label', 'io.hass.description=Converts Envoy Json to mqtt - V5, v7 and v8 with batteries', '--build-arg', 'BUILD_FROM=ghcr.io/home-assistant/amd64-base:latest', '--build-arg', 'BUILD_VERSION=1.0.22', '--build-arg', 'BUILD_ARCH=amd64']' on docker.io/library/docker:29.1.3-cli
2026-03-16 11:31:28.139 ERROR (MainThread) [supervisor.docker.addon] Docker build failed for 13e68335/amd64-addon-envoy_to_mqtt_json:1.0.22 (exit code 1). Build output:
#0 building with "default" instance using docker driver
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 508B done
#1 DONE 0.0s
#2 [internal] load metadata for ghcr.io/home-assistant/amd64-base:latest
#2 DONE 0.5s
#3 [internal] load .dockerignore
#3 transferring context: 701B done
#3 DONE 0.0s
#4 [1/6] FROM ghcr.io/home-assistant/amd64-base:latest@sha256:b4ccc4c08197c5bfee0d263758ab409f737383fb0140bdf034d2d481a8c964a8
#4 resolve ghcr.io/home-assistant/amd64-base:latest@sha256:b4ccc4c08197c5bfee0d263758ab409f737383fb0140bdf034d2d481a8c964a8 0.0s done
#4 DONE 0.0s
#5 [2/6] WORKDIR /app
#5 CACHED
#6 [internal] load build context
#6 transferring context: 23.30kB done
#6 DONE 0.0s
#7 [3/6] RUN apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3
#7 1.099 ( 1/19) Installing libbz2 (1.0.8-r6)
#7 1.106 ( 2/19) Installing libexpat (2.7.4-r0)
#7 1.113 ( 3/19) Installing libffi (3.5.2-r0)
#7 1.118 ( 4/19) Installing gdbm (1.26-r0)
#7 1.125 ( 5/19) Installing mpdecimal (4.0.1-r0)
#7 1.132 ( 6/19) Installing libpanelw (6.5_p20251123-r0)
#7 1.137 ( 7/19) Installing sqlite-libs (3.51.2-r0)
#7 1.203 ( 8/19) Installing python3 (3.12.12-r0)
#7 1.395 ( 9/19) Installing python3-pycache-pyc0 (3.12.12-r0)
#7 1.514 (10/19) Installing pyc (3.12.12-r0)
#7 1.514 (11/19) Installing py3-setuptools-pyc (80.9.0-r2)
#7 1.576 (12/19) Installing py3-pip-pyc (25.1.1-r1)
#7 1.651 (13/19) Installing py3-packaging-pyc (25.0-r0)
#7 1.660 (14/19) Installing python3-pyc (3.12.12-r0)
#7 1.660 (15/19) Installing py3-parsing (3.2.5-r0)
#7 1.669 (16/19) Installing py3-parsing-pyc (3.2.5-r0)
#7 1.681 (17/19) Installing py3-packaging (25.0-r0)
#7 1.690 (18/19) Installing py3-setuptools (80.9.0-r2)
#7 1.753 (19/19) Installing py3-pip (25.1.1-r1)
#7 1.823 Executing busybox-1.37.0-r30.trigger
#7 1.835 OK: 81.7 MiB in 70 packages
#7 2.268 error: externally-managed-environment
#7 2.268 
#7 2.268 × This environment is externally managed
#7 2.268 ╰─> 
#7 2.268     The system-wide python installation should be maintained using the system
#7 2.268     package manager (apk) only.
#7 2.268     
#7 2.268     If the package in question is not packaged already (and hence installable via
#7 2.268     "apk add py3-somepackage"), please consider installing it inside a virtual
#7 2.268     environment, e.g.:
#7 2.268     
#7 2.268     python3 -m venv /path/to/venv
#7 2.268     . /path/to/venv/bin/activate
#7 2.268     pip install mypackage
#7 2.268     
#7 2.268     To exit the virtual environment, run:
#7 2.268     
#7 2.268     deactivate
#7 2.268     
#7 2.268     The virtual environment is not deleted, and can be re-entered by re-sourcing
#7 2.268     the activate file.
#7 2.268     
#7 2.268     To automatically manage virtual environments, consider using pipx (from the
#7 2.268     pipx package).
#7 2.268 
#7 2.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
#7 2.268 hint: See PEP 668 for the detailed specification.
#7 ERROR: process "/bin/ash -o pipefail -c apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3" did not complete successfully: exit code: 1
------
 > [3/6] RUN apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3:
2.268     deactivate
2.268     
2.268     The virtual environment is not deleted, and can be re-entered by re-sourcing
2.268     the activate file.
2.268     
2.268     To automatically manage virtual environments, consider using pipx (from the
2.268     pipx package).
2.268 
2.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
2.268 hint: See PEP 668 for the detailed specification.
------
Dockerfile:8
--------------------
   7 |     # Install requirements
   8 | >>> RUN apk add --no-cache python3 py3-pip && \
   9 | >>>     pip install --no-cache-dir paho-mqtt requests urllib3
  10 |     
--------------------
ERROR: failed to build: failed to solve: process "/bin/ash -o pipefail -c apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3" did not complete successfully: exit code: 1
2026-03-16 11:31:28.139 ERROR (MainThread) [supervisor.addons.addon] Could not build image for addon 13e68335_envoy_to_mqtt_json: Docker build failed for 13e68335/amd64-addon-envoy_to_mqtt_json:1.0.22 (exit code 1). Build output:
#0 building with "default" instance using docker driver
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 508B done
#1 DONE 0.0s
#2 [internal] load metadata for ghcr.io/home-assistant/amd64-base:latest
#2 DONE 0.5s
#3 [internal] load .dockerignore
#3 transferring context: 701B done
#3 DONE 0.0s
#4 [1/6] FROM ghcr.io/home-assistant/amd64-base:latest@sha256:b4ccc4c08197c5bfee0d263758ab409f737383fb0140bdf034d2d481a8c964a8
#4 resolve ghcr.io/home-assistant/amd64-base:latest@sha256:b4ccc4c08197c5bfee0d263758ab409f737383fb0140bdf034d2d481a8c964a8 0.0s done
#4 DONE 0.0s
#5 [2/6] WORKDIR /app
#5 CACHED
#6 [internal] load build context
#6 transferring context: 23.30kB done
#6 DONE 0.0s
#7 [3/6] RUN apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3
#7 1.099 ( 1/19) Installing libbz2 (1.0.8-r6)
#7 1.106 ( 2/19) Installing libexpat (2.7.4-r0)
#7 1.113 ( 3/19) Installing libffi (3.5.2-r0)
#7 1.118 ( 4/19) Installing gdbm (1.26-r0)
#7 1.125 ( 5/19) Installing mpdecimal (4.0.1-r0)
#7 1.132 ( 6/19) Installing libpanelw (6.5_p20251123-r0)
#7 1.137 ( 7/19) Installing sqlite-libs (3.51.2-r0)
#7 1.203 ( 8/19) Installing python3 (3.12.12-r0)
#7 1.395 ( 9/19) Installing python3-pycache-pyc0 (3.12.12-r0)
#7 1.514 (10/19) Installing pyc (3.12.12-r0)
#7 1.514 (11/19) Installing py3-setuptools-pyc (80.9.0-r2)
#7 1.576 (12/19) Installing py3-pip-pyc (25.1.1-r1)
#7 1.651 (13/19) Installing py3-packaging-pyc (25.0-r0)
#7 1.660 (14/19) Installing python3-pyc (3.12.12-r0)
#7 1.660 (15/19) Installing py3-parsing (3.2.5-r0)
#7 1.669 (16/19) Installing py3-parsing-pyc (3.2.5-r0)
#7 1.681 (17/19) Installing py3-packaging (25.0-r0)
#7 1.690 (18/19) Installing py3-setuptools (80.9.0-r2)
#7 1.753 (19/19) Installing py3-pip (25.1.1-r1)
#7 1.823 Executing busybox-1.37.0-r30.trigger
#7 1.835 OK: 81.7 MiB in 70 packages
#7 2.268 error: externally-managed-environment
#7 2.268 
#7 2.268 × This environment is externally managed
#7 2.268 ╰─> 
#7 2.268     The system-wide python installation should be maintained using the system
#7 2.268     package manager (apk) only.
#7 2.268     
#7 2.268     If the package in question is not packaged already (and hence installable via
#7 2.268     "apk add py3-somepackage"), please consider installing it inside a virtual
#7 2.268     environment, e.g.:
#7 2.268     
#7 2.268     python3 -m venv /path/to/venv
#7 2.268     . /path/to/venv/bin/activate
#7 2.268     pip install mypackage
#7 2.268     
#7 2.268     To exit the virtual environment, run:
#7 2.268     
#7 2.268     deactivate
#7 2.268     
#7 2.268     The virtual environment is not deleted, and can be re-entered by re-sourcing
#7 2.268     the activate file.
#7 2.268     
#7 2.268     To automatically manage virtual environments, consider using pipx (from the
#7 2.268     pipx package).
#7 2.268 
#7 2.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
#7 2.268 hint: See PEP 668 for the detailed specification.
#7 ERROR: process "/bin/ash -o pipefail -c apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3" did not complete successfully: exit code: 1
------
 > [3/6] RUN apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3:
2.268     deactivate
2.268     
2.268     The virtual environment is not deleted, and can be re-entered by re-sourcing
2.268     the activate file.
2.268     
2.268     To automatically manage virtual environments, consider using pipx (from the
2.268     pipx package).
2.268 
2.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
2.268 hint: See PEP 668 for the detailed specification.
------
Dockerfile:8
--------------------
   7 |     # Install requirements
   8 | >>> RUN apk add --no-cache python3 py3-pip && \
   9 | >>>     pip install --no-cache-dir paho-mqtt requests urllib3
  10 |     
--------------------
ERROR: failed to build: failed to solve: process "/bin/ash -o pipefail -c apk add --no-cache python3 py3-pip &&     pip install --no-cache-dir paho-mqtt requests urllib3" did not complete successfully: exit code: 1

```

vk2him commented:
@JonathanGiles 
A quick google AI search on this indicates the method you're using may be attempting to update the entire Home Assistant Python environment instead of the environment within the addon container. It suggests this method instead:

How to Fix It
Option 1: Use Alpine's Pre-compiled Packages (Recommended)
This is the cleanest and fastest method for Home Assistant add-ons. It uses the system package manager (apk) instead of pip. This avoids compilation errors and complies with system restrictions. 

```
dockerfile
RUN apk add --no-cache \
    py3-paho-mqtt \
    py3-requests \
    py3-urllib3
```

Does this make sense to you? 
Thanks

sddgit commented:
FWIW, I fed your log to the AI I use, and it said:

## Issue

The Alpine Linux base image now ships Python 3.12, which enforces **PEP 668** — it blocks `pip install` from modifying the system-wide Python environment directly. You must either use a virtual environment or pass `--break-system-packages`.

## Fix

Update the `RUN` step in your `Dockerfile` using one of these two approaches:

### Option 1 — Virtual environment (recommended)

```dockerfile
RUN apk add --no-cache python3 py3-pip && \
    python3 -m venv /opt/venv && \
    /opt/venv/bin/pip install --no-cache-dir paho-mqtt requests urllib3

ENV PATH="/opt/venv/bin:$PATH"
```

### Option 2 — Override the restriction (simpler)

```dockerfile
RUN apk add --no-cache python3 py3-pip && \
    pip install --no-cache-dir --break-system-packages paho-mqtt requests urllib3
```

Option 1 is cleaner and avoids potential conflicts with system packages. If you go with Option 1, ensure your `CMD` or `ENTRYPOINT` uses the venv's Python, or rely on the `ENV PATH` line to pick it up automatically.

And no, I have no idea what this means! Hopefully @JonathanGiles can shed some light.
